dll/win32/browseui/regtreeoptions.cpp

```diff
@@ -1,59 +1,565 @@
+        {
+            if (SUCCEEDED(SHLoadRegUIStringW(hItemKey, L"Text", Temp, _countof(Temp))))
+                tvi.pszText = Temp;
+            else
+                tvi.pszText = Name;
+#ifdef __REACTOS__ // Grayed is a ROS extension
+            DWORD grayed = GetRegDWORD(hItemKey, L"Grayed", DWORD(FALSE));
+            if (grayed)
+            {
+                tvi.lParam = (LPARAM)HKEY_GRAYED;
+                hCloseKey = hItemKey;
+                tvi.iImage += 2; // Use disabled state image
+                if (grayed & 4)
+                    tvi.iImage = (tvi.iImage & ~1) | ((grayed & 2) != 0); // Force a specific state
+            }
+#endif
+            tvi.iSelectedImage = tvi.iImage;
+            tvis.hParent = hParent;
+            tvis.hInsertAfter = hInsertAfter;
+            Valid = (tvi.hItem = TreeView_InsertItem(m_hTree, &tvis)) != NULL;
+        }
+
+        if (Valid && Type == TYPE_GROUP)
+        {
+            AddItemsFromRegistry(hItemKey, tvi.hItem, TVI_FIRST);
+            TreeView_Expand(m_hTree, tvi.hItem, TVE_EXPAND);
+        }
+        if (!Valid)
+            RegCloseKey(hItemKey);
+        if (hCloseKey && hItemKey != hCloseKey)
+            RegCloseKey(hCloseKey);
+    }
+    SendMessage(m_hTree, TVM_SORTCHILDREN, 0, (LPARAM)hParent);
+}
+
+HRESULT STDMETHODCALLTYPE CRegTreeOptions::InitTree(HWND hTV, HKEY hKey, LPCSTR SubKey, char const *param18)
+{
+    UNREFERENCED_PARAMETER(param18);
+    m_hTree = hTV;
+
+    m_hIL = CreateStateImages(GetSystemMetrics(SM_CXSMICON));
+    HIMAGELIST hIL = TreeView_SetImageList(hTV, m_hIL, TVSIL_NORMAL);
+    if (hIL)
+        ImageList_Destroy(hIL);
+
+    WCHAR Path[MAX_PATH];
+    SHAnsiToUnicode(SubKey, Path, _countof(Path));
+    if (RegOpenKeyExW(hKey, Path, 0, KEY_ENUMERATE_SUB_KEYS, &hKey) == ERROR_SUCCESS)
+    {
+        AddItemsFromRegistry(hKey, TVI_ROOT, TVI_ROOT);
+        RegCloseKey(hKey);
+    }
+
+    TreeView_EnsureVisible(hTV, TreeView_GetRoot(hTV));
     return E_NOTIMPL;
 }
 
-HRESULT STDMETHODCALLTYPE CRegTreeOptions::GetSite(REFIID riid, void **ppvSite)
+HRESULT STDMETHODCALLTYPE CRegTreeOptions::WalkTree(WALK_TREE_CMD Command)
+{
+    for (HTREEITEM hTI = TreeView_GetRoot(m_hTree); hTI;)
+    {
+        ::WalkTree(Command, m_hTree, hTI);
+        hTI = TreeView_GetNextSibling(m_hTree, hTI);
+    }
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE CRegTreeOptions::ToggleItem(HTREEITEM hTI)
+{
+    TVITEM tvi;
+    tvi.mask = TVIF_PARAM | TVIF_IMAGE | TVIF_SELECTEDIMAGE;
+    tvi.hItem = hTI;
+    if (!tvi.hItem || !TreeView_GetItem(m_hTree, &tvi))
+        return E_INVALIDARG;
+
+    if (tvi.iImage >= STATE_GROUP || (HKEY)tvi.lParam == HKEY_GRAYED)
+        return S_FALSE;
+
+    if (tvi.iImage <= max(STATE_CHECKOFF, STATE_CHECKON))
+    {
+        tvi.iImage ^= 1;
+    }
+    else
+    {
+        HTREEITEM hRadio = TreeView_GetChild(m_hTree, TreeView_GetParent(m_hTree, hTI));
+        while (hRadio) // Turn off everyone in this group
```

Comment:
nit: 'toggle' instead of 'turn off'

Reply:
It does not toggle, it "unchecks" every radio button that is "checked". The clicked radio button is turned on later.

Reply:
Right, I missed the `tvi.iImage == STATE_RADIOON`.